Adding SemVer 2.0.0 support

emgarten opened the pull request:
This updates NuGet.SemanticVersion to support SemVer 2.0.0.

Primary change here is:
- Update the regex used to parse versions. These were previously used in NuGet.Versioning.
- Copied release label compare from NuGet.Versioning

Unit tests have been ported over from NuGet 3.x for version parsing and ranges.

Metadata will be carried along with the version but is only displayed when using one of the new methods, ToOriginalString, or ToFullString. This is the same behavior that NuGet.Versioning uses.

This change allows NuGet.Server and other projects still relying on NuGet.Core to handle SemVer 2.0.0.

Fixes https://github.com/NuGet/Home/issues/3383

//cc @xavierdecoster @joelverhagen @maartenba @rohit21agrawal @rrelyea @yishaigalatzer

emgarten committed:
Adding SemVer 2.0.0 support

xavierdecoster commented on src/Core/SemanticVersion.cs:
```diff
@@ -39,32 +45,49 @@ public SemanticVersion(int major, int minor, int build, string specialVersion)
         {
         }
 
+        public SemanticVersion(int major, int minor, int build, string specialVersion, string metadata)
```
In SemVer terms, the `build` argument is called `patch`. Perhaps we should use the same terminology from the SemVer.org spec to improve readability?

According to the spec, major, minor and patch numbers are mandatory, and labels for prerelease versions and build metadata are optional version parts. Knowing the SemVer spec, it is not clear to me what the `specialVersion` and `metadata` arguments refer to.

`Major.Minor.Patch-PrereleaseLabel+BuildMetadata`

emgarten replied:
I agree, but changing any of these would be a breaking change. The goal is here to allow the current code to read and sort SemVer 2.0.0 so that other projects can upgrade without making any other changes.

SpecialVersion is especially confusing, this class still treats versions as System.Version + a special pre-release label string since NuGet originally did not have pre-release.

I have made all of these changes in the past so that NuGet.Core.dll used a completely rewritten versioning class like NuGet.Version has. It took 1+ week of just fixing everywhere that SematicVersion was used along with the mock tests. The result was a change so large that it was never taken. This was all done ~2 years ago.

xavierdecoster replied:
Ouch! Perhaps adding a small comment in the code might clarify things a bit?

emgarten replied:
That sounds like a good idea.

xavierdecoster commented on src/Core/SemanticVersion.cs:
```diff
@@ -85,18 +108,27 @@ public string SpecialVersion
+        /// <summary>
+        /// SemVer 2.0.0 metadata. This is not used for comparing or sorting.
+        /// </summary>
+        public string Metadata
```
Assuming you refer to BuildMetadata?

emgarten replied:
Yes, it could be called that instead. NuGet.Versioning uses Metadata so I've added the same field here to match.

xavierdecoster commented:
Few nits about terminology used, but otherwise LGTM!

joelverhagen commented on src/Core/SemanticVersion.cs:
```diff
@@ -39,32 +45,49 @@ public SemanticVersion(int major, int minor, int build, string specialVersion)
-            _originalString = String.IsNullOrEmpty(originalString) ? version.ToString() + (!String.IsNullOrEmpty(specialVersion) ? '-' + specialVersion : null) : originalString;
+            Metadata = metadata;
+
+            _originalString = String.IsNullOrEmpty(originalString) ? version.ToString() 
```
If no original string is provided, we build one? That seems a little weird... can we leave original string null?

emgarten replied:
The original string is used for getting the version component inside the class, it may not be the greatest design, but rewriting the class to same a string seems unnecessary for this legacy code.

joelverhagen commented on src/Core/SemanticVersion.cs:
```diff
@@ -16,8 +17,13 @@ namespace NuGet
-        private static readonly Regex _strictSemanticVersionRegex = new Regex(@"^(?<Version>\d+(\.\d+){2})(?<Release>-[a-z][0-9a-z-]*)?$", _flags);
+
+        // Versions containing up to 4 digits
+        private static readonly Regex _semanticVersionRegex = new Regex(@"^(?<Version>\d+(\s*\.\s*\d+){0,3})(?<Release>-([0]\b|[0]$|[0][0-9]*[A-Za-z-]+|[1-9A-Za-z-][0-9A-Za-z-]*)+(\.([0]\b|[0]$|[0][0-9]*[A-Za-z-]+|[1-9A-Za-z-][0-9A-Za-z-]*)+)*)?(?<Metadata>\+[0-9A-Za-z-]+(\.[0-9A-Za-z-]+)*)?$", _flags);
```
This regex is a monster. Would it be possible to move it to a string const, break it up into multiple lines, and comment? There are definitely some repeated substrings. Could those maybe me moved to little descriptive constants?

I can follow almost all of the regex, but what does `[0]\b` do? That looks like a one character character class, which is odd.

emgarten replied:
It refers to a leading zero here which is not allowed for labels without letters.

This regex matches the semver spec, users do not need to reverse engineer the regex to find the rules here, they aren't nuget specific.

joelverhagen commented on src/Core/SemanticVersion.cs:
```diff
@@ -16,8 +17,13 @@ namespace NuGet
     public sealed class SemanticVersion : IComparable, IComparable<SemanticVersion>, IEquatable<SemanticVersion>
     {
         private const RegexOptions _flags = RegexOptions.Compiled | RegexOptions.IgnoreCase | RegexOptions.ExplicitCapture;
```
If we are doing `IgnoreCase`, we don't have to have `A-Za-z`.

emgarten replied:
True, that flag could be removed now.

joelverhagen commented:
:watch:

emgarten committed:
Removing regex IgnoreCase flag now that it is supported in the regex itself

emgarten commented:
Updated, I've removed the IgnoreCase flag. The rest of the feedback is valid, but the goal here is to avoid all breaking API changes, and to not to rewrite the SemanticVersion class. Any users needing that should move to NuGet.Versioning.

Updating NuGet.Core is intended to help parts of the eco system still on NuGet.Core.dll to get SemVer 2.0.0 support without changing anything. They should move to NuGet.Versioning and the 3.x libraries when possible.

joelverhagen commented:
Sure, sounds fine. :shipit:

emgarten commented:
I've added more XML comments to help with the SemVer vs random NuGet.SemanticVersion names for things.

This will be in 2.14
